airflow/jobs/triggerer_job.py

```diff
@@ -315,6 +339,7 @@ async def cleanup_finished_triggers(self):
                         details["name"],
                     )
                     self.failed_triggers.append(trigger_id)
+                self.completed_triggers.add(trigger_id)
```

Comment:
It appears that `completed_triggers` will grow in an unbounded fashion with this code over time - I would suggest a datastructure that is self-limiting. My initial idea is a dict with the IDs as keys and the datetime they were inserted as values, and then just remove all keys whose values are more than 5 minutes old in the first part of `purge_completed_triggers`.

Reply:
how do you envision it growing?  the items should only be in `completed_triggers` for basically one or two loop cycles. it goes `submit_event` (into completed) -> `Trigger.clean_unused` (purged from db) -> `purge_completed_triggers_list` (purged from completed)

but maybe you're saying that in the wild the code may not behave as intended, and triggers will accumulate in the DB, and therefore they'll accumulate in the `completed_triggers` set?  but if they accumulate in the db they'll keep getting recreated and eventually that would be an issue in itself because not only would they exist as ids in a set but they would be running.

using a TTL approach as you have suggested would avoid a db query though.

LMK your thoughts.

Reply:
@andrewgodwin gentle nudge here.  If you want to go with time-based expiration I have no objection and am happy to rework this.  I do recognize that my solution adds a query but if you don't mind explaining I'm having trouble seeingthe unbound growth.

Reply:
I guess I just don't quite trust `difference_update` quite enough, I missed it in my first read-through for "what in here cleans out the datastructure?"

If you're happy there's no edge cases where it can leak entries and write a test to prove so, I think my objection here is void.